Sources/PenguinStructures/ArrayStorage.swift

```diff
@@ -226,18 +272,39 @@ extension ArrayStorageImplementation {
     }
   }
 
-  /// Returns an empty instance with `capacity` at least `minimumCapacity`.
-  public static func create(minimumCapacity: Int) -> Self {
-    unsafeDowncast(
-      Accessor(
-        bufferClass: Self.self, minimumCapacity: minimumCapacity
-      ) { buffer, getCapacity in 
-        ArrayHeader(count: 0, capacity: getCapacity(buffer))
-      }.buffer,
-      to: Self.self)
+  /// Creates an empty instance with `capacity` at least `minimumCapacity`.
+  public init(minimumCapacity: Int) {
+    let access = Accessor(
+      bufferClass: Self.self, minimumCapacity: minimumCapacity
+    ) { buffer, getCapacity in 
+      ArrayHeader(count: 0, capacity: getCapacity(buffer))
+    }
+    
+    self.init(
+      unsafelyAliasing: unsafeDowncast(access.buffer, to: FactoryBase.self))
+  }
+  
+  /// Creates an instance with the given `count`, and capacity at least
+  /// `minimumCapacity`, and elements initialized by `initializeElements`,
+  /// which is passed the address of the (uninitialized) first element.
+  ///
+  /// - Requires: `initializeElements` initializes exactly `count` contiguous
+  ///   elements starting with the address it is passed.
+  public init(
+    count: Int,
+    minimumCapacity: Int = 0,
+    initializeElements: (_ baseAddress: UnsafeMutablePointer<Element>) -> Void
```

Comment:
Should `initializeElements` be given `count` as an additional parameter? Or alternatively, should it be given an `UnsafeMutableBufferPointer` instead?

Reply:
In fact, no.  The caller has the count and can easily bind it into the closure, and because I couldn't get the fix past one person's objections, `Unsafe[Mutable]BufferPointer` still can have a `nil` base address, which is just a horrible inconvenience and potential efficiency cost.

At some point we should make an improved `Unsafe[Mutable]BufferPointer` that doesn't have this problem.  Keep in mind that the elements would be uninitialized, so any use of the collection conformance would be UB.

There's a different version of this that supports construction from sequences of unknown length, and for that the actual count needs to be an output of the closure, but I don't need that here and don't really care about single-pass things today.

Reply:
Filed https://github.com/saeta/penguin/issues/68 for follow-up on improved `Unsafe[Mutable]BufferPointer`. I had always wondered wby `baseAddress` could be `nil`; it's always been annoying. Glad I wasn't missing something big. :-)

Reply:
Ask me about why offline ;-)